src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,1062 @@
+        x = F.dropout(x, p=self.activation_dropout, training=self.training)
+        x = self.fc2(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        x = residual + x
+        x = self.final_layer_norm(x)
+        return (
+            x,
+            self_attn_weights,
+            cached_states,
+        )  # just self_attn weights for now, following t5, cached_states = cache for decoding
+
+    def _past_to_dict(self, prev_attn_state):
+        prev_key, prev_value = prev_attn_state[:2]
+        saved_state = {"prev_key": prev_key, "prev_value": prev_value}
+        if len(prev_attn_state) >= 3:
+            saved_state["prev_key_padding_mask"] = prev_attn_state[2]
+        return saved_state
+
+
+class BartDecoder(nn.Module):
+    """
+    Transformer decoder consisting of *config.decoder_layers* layers. Each layer
+    is a :class:`DecoderLayer`.
+    Args:
+        config: BartConfig
+        embed_tokens (torch.nn.Embedding): output embedding
+    """
+
+    def __init__(self, config: BartConfig, embed_tokens: nn.Embedding):
+        super().__init__()
+        self.output_past = config.output_past
+        self.output_attentions = config.output_attentions
+        self.output_hidden_states = config.output_hidden_states
+        self.dropout = config.dropout
+        self.layerdrop = config.decoder_layerdrop
+        self.padding_idx = embed_tokens.padding_idx
+        self.max_target_positions = config.max_position_embeddings
+        self.embed_tokens = embed_tokens
+        self.embed_positions = LearnedPositionalEmbedding(
+            config.max_position_embeddings, config.d_model, self.padding_idx,
+        )
+        self.layers = nn.ModuleList(
+            [DecoderLayer(config) for _ in range(config.decoder_layers)]
+        )  # type: List[DecoderLayer]
+        self.layernorm_embedding = LayerNorm(config.d_model)
+
+    def forward(
+        self, input_ids, encoder_hidden_states, encoder_padding_mask, combined_mask, cached_states=None, **unused
+    ):
+        """
+        Includes several features from "Jointly Learning to Align and
+        Translate with Transformer Models" (Garg et al., EMNLP 2019).
+
+        Args:
+            input_ids (LongTensor): previous decoder outputs of shape
+                `(batch, tgt_len)`, for teacher forcing
+            encoder_hidden_states: output from the encoder, used for
+                encoder-side attention
+            encoder_padding_mask: for ignoring pad tokens
+            cached_states (dict or None): dictionary used for storing state during generation
+
+        Returns:
+            tuple:
+                - the decoder's features of shape `(batch, tgt_len, embed_dim)`
+                - hidden states
+                - attentions
+        """
+        # embed positions
+        positions = self.embed_positions(input_ids, cached_states=cached_states)
+
+        if cached_states is not None:
+            input_ids = input_ids[:, -1:]
```

Comment:
That's great - it should make the `prepare_inputs_for_generation` fn much easier since the `input_ids` don't have to be cut when the cached_states are used as an input (as has to be done in GPT-2).